File: Afonso/Serie1/test/BoundedLazyTest.java
import main.boundedLazy.BoundedLazy;
import org.junit.jupiter.api.Test;

import java.util.Optional;
import java.util.function.Supplier;

import static org.junit.jupiter.api.Assertions.assertEquals;

class BoundedLazyTest {

    private long TIMEOUT = 2000L;

    private final int NR_OF_THREADS = 5;
    Thread[] threads = new Thread[NR_OF_THREADS];

    private final Supplier<Long> supplier = () -> Thread.currentThread().getId();

    @Test
    void shouldReturnValueAlreadyCreated() {
        int lives = 3;
        BoundedLazy<Long> boundedLazy = new BoundedLazy<>(supplier, lives);
        Optional[] value = {Optional.of(Long.MAX_VALUE)};

        createThreads(2, () -> {
            try {
                value[0] = boundedLazy.get(TIMEOUT);
            } catch (Exception ignored) {
            }
        });

        threads[0].start();
        try {
            threads[0].join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        threads[1].start();
        try {
            threads[1].join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        assertEquals(threads[0].getId(), (Long) value[0].get());
        clearThreadArray();
    }

    @Test
    void shouldExceedNumberOfLivesAndComputeValueAgain() {
        BoundedLazy<Long> boundedLazy = new BoundedLazy<>(supplier, 2);
        Optional[] values = {
                Optional.of(Long.MAX_VALUE),
                Optional.of(Long.MAX_VALUE),
                Optional.of(Long.MAX_VALUE)
        };
        int [] idx = {0};
        createThreads(3, () -> {
            try {
                values[idx[0]++] = boundedLazy.get(1000);
            } catch (Exception ignored) {
            }
        });

        for (int i = 0; i < 3; i++) {
            threads[i].start();
            try {
                threads[i].join();
            } catch (InterruptedException ignored) {
            }
        }

        assertEquals(threads[0].getId(), (Long) values[0].get());
        assertEquals(threads[0].getId(), (Long) values[1].get());
        assertEquals(threads[2].getId(), (Long) values[2].get());
        clearThreadArray();
    }

    private void createThreads(int nr, Runnable runnable) {
        for (int i = 0; i < nr; i++) {
            threads[i] = new Thread(runnable);
            threads[i].setPriority(Thread.MAX_PRIORITY);
        }
    }

    private void clearThreadArray() {
        threads = new Thread[NR_OF_THREADS];
    }

    void soutThreadRunningTime(int threadIdx, long totalTime) {
        System.out.println("threads[" + threadIdx + "] total time running: " +
                                   (totalTime > 1000000000 ?
                                           (float) (totalTime / 1000000000) + " seconds" :
                                           totalTime > 1000000 ?
                                                   (float) (totalTime / 1000000) + " miliseconds" :
                                                   (totalTime / 1000) + " microsenconds"
                                   ));
    }
}

File: Afonso/Serie1 .NET/.NET/Serie1/Serie1Test/BoundedLazyTest.cs
using System;
using System.Threading;
using Ex1;
using NUnit.Framework;
using Utils;
using static Serie1Test.TestUtils;
using Assert = Microsoft.VisualStudio.TestTools.UnitTesting.Assert;

namespace Serie1Test {
    class BoundedLazyTest {
        /**
         * DEBUG ONLY VARIABLES
         */
        private readonly string _testClass = "TEST CLASS - ";

        /**
         * End of DEBUG ONLY VARIABLES
         */
        private readonly int TIMEOUT = 2000;

        private const int NrOfThreads = 5;
        Thread[] _threads = new Thread[NrOfThreads];

        private readonly Func<long> _supplier = () => Thread.CurrentThread.ManagedThreadId;

        [Test]
        public void ShouldReturnValueAlreadyCreated() {
            int lives = 3;
            BoundedLazy<long> boundedLazy = new BoundedLazy<long>(_supplier, lives);
            Optional<long>[] value = {Optional<long>.Of(long.MaxValue)};

            CreateThreads(_threads, 2, () => {
                try {
                    value[0] = boundedLazy.Get(TIMEOUT);
                }
                catch (Exception) { }
            });

            _threads[0].Start();
            try {
                _threads[0].Join();
            }
            catch (ThreadInterruptedException e) {
                Console.WriteLine(e.StackTrace);
            }

            _threads[1].Start();
            try {
                _threads[1].Join();
            }
            catch (ThreadInterruptedException e) {
                Console.WriteLine(e.StackTrace);
            }

            Assert.AreEqual(_threads[0].ManagedThreadId, value[0].Value);
            _threads = ClearThreadArray(NrOfThreads);
        }

        [Test]
        public void ShouldExceedNumberOfLivesAndComputeValueAgain() {
            BoundedLazy<long> boundedLazy = new BoundedLazy<long>(_supplier, 2);
            Optional<long>[] values = {
                Optional<long>.Of(long.MaxValue),
                Optional<long>.Of(long.MaxValue),
                Optional<long>.Of(long.MaxValue)
            };
            int[] idx = {0};
            CreateThreads(_threads, 3, () => {
                try {
                    values[idx[0]++] = boundedLazy.Get(1000);
                }
                catch (Exception) { }
            });

            for (int i = 0; i < 3; i++) {
                _threads[i].Start();
                try {
                    _threads[i].Join();
                }
                catch (ThreadInterruptedException) { }
            }

            Assert.AreEqual(_threads[0].ManagedThreadId, (long) values[0].Value);
            Assert.AreEqual(_threads[0].ManagedThreadId, (long) values[1].Value);
            Assert.AreEqual(_threads[2].ManagedThreadId, (long) values[2].Value);
            _threads = ClearThreadArray(NrOfThreads);
        }

        [Test]
        public void ShouldReturnInterruptedException() {
            BoundedLazy<long> boundedLazy = new BoundedLazy<long>(() => {
                try {
                    Thread.Sleep(2000);
                }
                catch (Exception ignored) { }

                return 5L;
            }, 2);

            bool expected = false;
            _threads[0] = new Thread(() => {
                try {
                    boundedLazy.Get(TIMEOUT);
                }
                catch (Exception ignored) { }
            });
            _threads[1] = new Thread(() => {
                try {
                    Thread.Sleep(1000);
                    boundedLazy.Get(TIMEOUT);
                }
                catch (ThreadInterruptedException ignored) {
                    expected = true;
                }
                catch (Exception ignored) { }
            });

            _threads[0].Start();
            _threads[1].Start();

            _threads[1].Interrupt();

            for (int i = 0; i < 2; i++) {
                try {
                    _threads[i].Join();
                }
                catch (ThreadInterruptedException ignored) { }
            }

            Assert.IsTrue(expected);
        }
    }
}